rccheck/src/lib.rs

```diff
@@ -0,0 +1,232 @@
+    ) -> Result<ClientCertVerified, rustls::Error> {
+        // Check this matches the key we expect
+        let cert = X509Certificate::from_der(&end_entity.0[..])
+            .map_err(|_| rustls::Error::InvalidCertificateEncoding)?;
+        let spki = cert.1.public_key().clone();
+        if spki != self.0 {
+            return Err(rustls::Error::InvalidCertificateData(format!(
+                "invalid peer certificate: received {:?} instead of expected {:?}",
+                spki, self.0
+            )));
+        }
+
+        // We now check we're receiving correctly signed data with the expected key
+        let (cert, chain, trustroots) = prepare_for_self_signed(end_entity, intermediates)?;
+        let now = webpki::Time::try_from(now).map_err(|_| rustls::Error::FailedToGetCurrentTime)?;
+        cert.verify_is_valid_tls_client_cert(
+            SUPPORTED_SIG_ALGS,
+            &webpki::TlsClientTrustAnchors(&trustroots),
+            &chain,
+            now,
+        )
+        .map_err(pki_error)
+        .map(|_| ClientCertVerified::assertion())
+    }
+}
+
+impl<'a> ServerCertVerifier for Psk<'a> {
+    fn verify_server_cert(
+        &self,
+        end_entity: &rustls::Certificate,
+        intermediates: &[rustls::Certificate],
+        server_name: &rustls::ServerName,
+        scts: &mut dyn Iterator<Item = &[u8]>,
+        ocsp_response: &[u8],
+        now: std::time::SystemTime,
+    ) -> Result<rustls::client::ServerCertVerified, rustls::Error> {
+        // Check this matches the key we expect
+        let cert = X509Certificate::from_der(&end_entity.0[..])
+            .map_err(|_| rustls::Error::InvalidCertificateEncoding)?;
+        let spki = cert.1.public_key().clone();
+        if spki != self.0 {
+            return Err(rustls::Error::InvalidCertificateData(format!(
+                "invalid peer certificate: received {:?} instead of expected {:?}",
+                spki, self.0
+            )));
+        }
+
+        // Then we check this is actually a valid self-signed certificate with matching name
+        let (cert, chain, trustroots) = prepare_for_self_signed(end_entity, intermediates)?;
+        let webpki_now =
+            webpki::Time::try_from(now).map_err(|_| rustls::Error::FailedToGetCurrentTime)?;
+
+        let dns_nameref = match server_name {
+            rustls::ServerName::DnsName(dns_name) => {
+                webpki::DnsNameRef::try_from_ascii_str(dns_name.as_ref())
+                    .map_err(|_| rustls::Error::UnsupportedNameType)?
+            }
+            _ => return Err(rustls::Error::UnsupportedNameType),
+        };
+
+        let cert = cert
+            .verify_is_valid_tls_server_cert(
+                SUPPORTED_SIG_ALGS,
+                &webpki::TlsServerTrustAnchors(&trustroots),
+                &chain,
+                webpki_now,
+            )
+            .map_err(pki_error)
+            .map(|_| cert)?;
+
+        let mut peekable = scts.peekable();
```

Comment:
Here we seem to be logging missing info, but not aborting / returning Err. What is the nature of these checks?

Reply:
This is the certificate transparency extension of TLS, which we can't do anything clever about, since we are actually bypassing the certificate chain mechanism (we are only using TLS as a vessel for checking a public key against our knowledge + a PoP).